sky/provision/oci/instance.py

```diff
@@ -0,0 +1,455 @@
+    for running_instance in running_instances:
+        inst = _get_inst_obj_with_ip(region, running_instance)
+        instances[inst['id']] = [
+            common.InstanceInfo(
+                instance_id=inst['id'],
+                internal_ip=inst['internal_ip'],
+                external_ip=inst['external_ip'],
+                tags=inst['tags'],
+            )
+        ]
+
+    instances = dict(sorted(instances.items(), key=lambda x: x[0]))
+    logger.debug(f'Cluster info: {instances}')
+
+    head_instance_id = _get_head_instance_id(running_instances)
+    logger.debug(f'Head instance id is {head_instance_id}')
+
+    return common.ClusterInfo(
+        provider_name='oci',
+        head_instance_id=head_instance_id,
+        instances=instances,
+        provider_config=provider_config,
+    )
+
+
+def _get_filtered_nodes(region, tag_filters):
+    return_nodes = []
+
+    try:
+        insts = query_helper.query_instances_by_tags(tag_filters, region)
+    except oci_adaptor.oci.exceptions.ServiceError as e:
+        with ux_utils.print_exception_no_traceback():
+            raise exceptions.ClusterStatusFetchingError(
+                f'Failed to query status for OCI cluster {tag_filters}.'
+                'Details: '
+                f'{common_utils.format_exception(e, use_bracket=True)}')
+
+    for inst in insts:
+        inst_id = inst.identifier
+        return_nodes.append({
+            'inst_id': inst_id,
+            'name': inst.display_name,
+            'ad': inst.availability_domain,
+            'compartment': inst.compartment_id,
+            'status': inst.lifecycle_state,
+            'oci_tags': inst.freeform_tags,
+        })
+
+    return [
+        v for v in return_nodes
+        if v['status'] in ('PROVISIONING', 'STARTING', 'RUNNING', 'STOPPED',
+                           'STOPPING')
+    ]
+
+
+def _get_inst_obj_with_ip(region, inst_info):
+    list_vnic_attachments_response = oci_adaptor.get_core_client(
+        region, oci_utils.oci_config.get_profile()).list_vnic_attachments(
+            availability_domain=inst_info['ad'],
+            compartment_id=inst_info['compartment'],
+            instance_id=inst_info['inst_id'],
+        )
+
+    vnic = list_vnic_attachments_response.data[0]
+    get_vnic_response = (oci_adaptor.get_net_client(
+        region,
+        oci_utils.oci_config.get_profile()).get_vnic(vnic_id=vnic.vnic_id).data)
+
+    internal_ip = get_vnic_response.private_ip
+    external_ip = get_vnic_response.public_ip
+    if external_ip is None:
+        external_ip = internal_ip
+
+    return {
+        'id': inst_info['inst_id'],
+        'name': inst_info['name'],
+        'external_ip': external_ip,
+        'internal_ip': internal_ip,
+        'tags': inst_info['oci_tags'],
+        'status': inst_info['status'],
+    }
+
+
+def _wait_until_status(region, node_id, status):
+    compute_client = oci_adaptor.get_core_client(
+        region, oci_utils.oci_config.get_profile())
```

Comment:
Seems we are constructing a lot of clients in those functions. Should we consider implement a similar function with cache like this?

https://github.com/skypilot-org/skypilot/blob/d3be8ed56adfb5efb9789790bec28616d446c239/sky/adaptors/azure.py#L83-L90